common/src/main/java/io/netty/util/NettyRuntime.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.util;
+
+import io.netty.util.internal.SystemPropertyUtil;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A utility class for wrapping calls to {@link Runtime}.
+ */
+public final class NettyRuntime {
+
+    private static final AtomicInteger availableProcessors = new AtomicInteger();
+
+    /**
+     * No public constructor to prevent instances from being created.
+     */
+    private NettyRuntime() {
+    }
+
+    /**
+     * Set the number of available processors.
+     *
+     * @param availableProcessors the number of available processors
+     * @throws IllegalArgumentException if the specified number of available processors is non-positive
+     * @throws IllegalStateException    if the number of available processors is already configured
+     */
+    public static void setAvailableProcessors(final int availableProcessors) {
+        if (availableProcessors <= 0) {
+            throw new IllegalArgumentException(
+                    "availableProcessors must be positive but was [" + availableProcessors + "]");
+        }
+        if (!NettyRuntime.availableProcessors.compareAndSet(0, availableProcessors)) {
+            throw new IllegalStateException(
+                    "availableProcessors is already set to [" + NettyRuntime.availableProcessors.get() + "], " +
+                            "rejecting [" + availableProcessors + "]");
+        }
+    }
+
+    /**
+     * Get the configured number of available processors. The default is {@link Runtime#availableProcessors()}. This
+     * can be overridden by setting the system property "io.netty.availableProcessors" or by invoking
+     * {@link #setAvailableProcessors(int)} before any calls to {@link #availableProcessors()}.
+     *
+     * @return the configured number of available processors
+     */
+    @SuppressForbidden(reason = "to obtain default number of available processors")
+    public static int availableProcessors() {
+        if (NettyRuntime.availableProcessors.get() == 0) {
+            final int availableProcessors =
+                    SystemPropertyUtil.getInt(
+                            "io.netty.availableProcessors",
+                            Runtime.getRuntime().availableProcessors());
+            setAvailableProcessors(availableProcessors);
```

Comment:
> I've acknowledged since the beginning that a get race is possible, we're just arguing about how to handle it.

The following seems like reasonable behavior to me:
- OK - Two threads race to initialize from `System.getProperty` and result in setting `availableProcessors` to the same value
- ERROR - Two threads race between initialize from `System.getProperty` and `setAvailableProcessors`
- ERROR - Two threads race between `setAvailableProcessors`
- ERROR - No race, but attempting to call `setAvailableProcessors` after `availableProcessors` variable has been set

Based upon the discussion it seems like you disagree and think that the first case should also be an ERROR with the rational that "this shouldn't happen". Is this correct?

> this might not be the case if one racing thread is getting before another racing thread programmatically sets the system property.

My suggestion acknowledged this and proposes and alternative approach to differentiate between the two scenarios. The point is this condition seems plausible. I have not seen a convincing argument that we should ignore this case.